Unit test create

guewa opened the pull request:
**What this PR does / why we need it**:
Unit test for command 'component-archive create'
**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Release note**:
<!--  Write your release note:
1. Enter your release note in the below block.
2. If no release note is required, just write "NONE" within the block.

Format of block header: <category> <target_group>
Possible values:
- category:       breaking|feature|bugfix|doc|other
- target_group:   user|operator|developer|dependency
-->
```other developer

```

guewa committed:
Unit test for create

guewa committed:
UnitTest for create command

achimweigel reviewed (approved):
/lgtm

schrodit commented on pkg/commands/componentarchive/create_test.go:
```diff
@@ -0,0 +1,58 @@
+		opts.Version = "v0.0.1"
+		opts.ComponentArchivePath = "./create-test"
+
+		fileName := "component-descriptor.yaml"
```
you should use the constant from the component-spec instead: https://github.com/gardener/component-spec/blob/master/bindings-go/ctf/ctf.go#L32

guewa replied:
ok, will be fixed

schrodit commented on pkg/commands/componentarchive/create_test.go:
```diff
@@ -0,0 +1,58 @@
+		cd := &cdv2.ComponentDescriptor{}
+		Expect(codec.Decode(data, cd)).To(Succeed())
+		Expect(cd.Name).To(Equal(opts.Name), "component name should be the same")
+		Expect(cd.Version).To(Equal(opts.Version), "component version should be the same")
```
we should also check the repository context

guewa replied:
done

schrodit reviewed (changes requested):
please check the pipeline there are some unformatted files. (just run `make format` or better `make all`)

guewa committed:
Unit test create

guewa committed:
Unit test create

schrodit commented on pkg/commands/componentarchive/create_test.go:
```diff
@@ -0,0 +1,83 @@
+		Expect(cd.Name).To(Equal(opts.Name), "component name should be the same")
+		Expect(cd.Version).To(Equal(opts.Version), "component version should be the same")
+
+		Expect(len(cd.RepositoryContexts) > 0).To(BeTrue(), "The repository contexts should return some data")
+		repoCtx := cd.RepositoryContexts[0]
+		Expect(repoCtx.GetType()).To(Equal(cdv2.OCIRegistryType), "check the repository context")
```
we should also check that the base url was correctly set like
```suggestion
		Expect(len(cd.RepositoryContexts) > 0).To(BeTrue(), "The repository contexts should return some data")
		repoCtx := cd.RepositoryContexts[0]
		Expect(repoCtx.GetType()).To(Equal(cdv2.OCIRegistryType), "check the repository context")
		ociRepoCtx := &cdv2.OCIRegistry{}
		Expect(repoCtx.DecodeInto(ociRepoCtx)).To(Succeed)
		Expect(ociRepoCtx.BaseUrl).To(Equal("example.com/testurl"))
```

schrodit reviewed (dismissed):
there are still linter issues

guewa committed:
Unit test ca create (linter fix)

Update pkg/commands/componentarchive/create_test.go

check base url of repository context

Unit test create

Unit test create

	modified:   pkg/commands/componentarchive/create_test.go

Co-Authored-By: Tim Schrodi <tschrodi96@googlemail.com>

guewa committed:
unit test create after make format

schrodit commented on pkg/commands/componentarchive/create_test.go:
```diff
@@ -0,0 +1,88 @@
+		opts.Name = "example.com/component/name"
+		opts.Version = "v0.0.1"
+		opts.ComponentArchivePath = "./create-test"
+		//opts.ComponentNameMapping = "urlPath"
```
```suggestion
```

schrodit commented on pkg/commands/componentarchive/create_test.go:
```diff
@@ -0,0 +1,88 @@
+		Expect(cd.Name).To(Equal(opts.Name), "component name should be the same")
+		Expect(cd.Version).To(Equal(opts.Version), "component version should be the same")
+
+		Expect(len(cd.RepositoryContexts) > 0).To(BeTrue(), "The repository contexts should return some data")
```
There should exactly one repository context
```suggestion
		Expect(cd.RepositoryContexts).To(HaveLen(1), "The repository contexts should return some data")
```

schrodit commented on pkg/commands/componentarchive/create_test.go:
```diff
@@ -0,0 +1,88 @@
+		Expect(repoCtx.DecodeInto(ociRepoCtx)).To(Succeed())
+		Expect(ociRepoCtx.BaseURL).To(Equal(opts.BaseUrl))
+
+		// check overwrite
+		opts.Version = "v0.0.2"
+		Expect(opts.Run(context.TODO(), logr.Discard(), testdataFs)).To(HaveOccurred(), "Should not overwrite existing component")
+
+		opts.Overwrite = true
+		Expect(opts.Run(context.TODO(), logr.Discard(), testdataFs)).To(Succeed(), "Should overwrite existing component")
+
+		data, err = vfs.ReadFile(testdataFs, filepath.Join(opts.ComponentArchivePath, ctf.ComponentDescriptorFileName))
+		Expect(err).ToNot(HaveOccurred())
+
+		cd = &cdv2.ComponentDescriptor{}
+		Expect(codec.Decode(data, cd)).To(Succeed())
+		Expect(cd.Name).To(Equal(opts.Name), "component name should be the same")
+		Expect(cd.Version).To(Equal(opts.Version), "component version should be the same")
+
+		Expect(len(cd.RepositoryContexts) > 0).To(BeTrue(), "The repository contexts should return some data")
+		repoCtx = cd.RepositoryContexts[0]
+		Expect(repoCtx.GetType()).To(Equal(cdv2.OCIRegistryType), "check the repository context")
+		Expect(repoCtx.DecodeInto(ociRepoCtx)).To(Succeed())
+		Expect(ociRepoCtx.BaseURL).To(Equal(opts.BaseUrl))
```
can we have that in separate test so that we test one functionality per feature which is easier to debug in case on issues

guewa replied:
all done

guewa committed:
Update unit test create

schrodit commented on pkg/commands/componentarchive/create_test.go:
```diff
@@ -0,0 +1,120 @@
+			Expect(cd.Name).To(Equal(opts.Name), "component name should be the same")
+			Expect(cd.Version).To(Equal(opts.Version), "component version should be the same")
+
+			Expect(len(cd.RepositoryContexts) > 0).To(BeTrue(), "The repository contexts should return some data")
```
```suggestion
			Expect(cd.RepositoryContexts).To(HaveLen(1), "The repository contexts should return some data")
```

guewa replied:
sorry, forgot this one

guewa committed:
update unit test create (only 1 repo ctx)

schrodit reviewed (approved):
/lgtm
